### DISCUSSION TITLE
How to register a custom router in GLSP?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1440] [Q&A] [answered]
Hi,
can someone explain me how I can register my own Router?
Some time ago we had [this discussion](https://github.com/eclipse-glsp/glsp/discussions/552). But the example seems no longer to be valid.

My router looks like this:

```javascript
import {
    GLSPManhattanEdgeRouter,
    GRoutableElement,
    ResolvedHandleMove
} from '@eclipse-glsp/client';

export class BPMNManhattanRouter extends GLSPManhattanEdgeRouter {

    static readonly BPMNKIND = 'bpmnmanhattan';

    override get kind() {
        return BPMNManhattanRouter.BPMNKIND;
    }
    protected override applyInnerHandleMoves(edge: GRoutableElement, moves: ResolvedHandleMove[]): void {
        super.applyInnerHandleMoves(edge, moves);


        console.log('├── New Routing Points:');
        edge.routingPoints.forEach(point => {
            console.log('│   ├── x=' + point.x + ' y=' + point.y);
        });

    }
}
```

And I want to register it in my DiagramModule like this:

```javascript
    bind(BPMNManhattanRouter).toSelf().inSingletonScope();
    bind(TYPES.IEdgeRouter).toService(BPMNManhattanRouter);
```

But during runtime the router kind is not know :-( I go the following error:

```
registry.ts:128 Uncaught (in promise) Error: Unknown registry key: bpmnmanhattan:rectangular
    at AnchorComputerRegistry.missing (registry.ts:128:15)
    at AnchorComputerRegistry.get (registry.ts:123:25)
    at AnchorComputerRegistry.get (anchor.ts:61:22)
    at BPMNManhattanRouter.getAnchorComputer (abstract-edge-router.ts:211:35)
    at BPMNManhattanRouter.getTranslatedAnchor (abstract-edge-router.ts:204:37)
    at BPMNManhattanRouter.route (manhattan-edge-router.ts:50:35)
    at EdgeRouterRegistry.route (routing.ts:234:23)
    at routing.ts:211:47
    at Map.forEach (<anonymous>)
    at EdgeRouterRegistry.doRouteAllChildren (routing.ts:205:22)
```

I guess my way to register the router is wrong?

**Comment by tortmayr** [comment]
@rsoika No your way to register the router is correct. The problem is since you have introduced a new router kind you also need to configure`IAnchorComputer`s for this new router kind. (see also https://github.com/eclipse-glsp/glsp/discussions/1435#discussioncomment-11478462).

Is there a specific reason that you want to introduce an entirely new routerkind? Do you plan on supporting both default `manhattan` and `bpmnmanhattan` in your application? If that is not the case I would recommend to directly customize the `manhattan` router by rebinding the `GLSPManhattanRouter` to your custom class: 
```ts
rebind(GLSPManhattanRouter).to(BPMNManhattanRouter).inSIngletonScope();
```
This way you don't have to concern yourself with configuring anchor computers since they are already configured [by default](https://github.com/eclipse-glsp/glsp-client/blob/dd011eae4f7a629cc2431e5e6d832ddb9483554b/packages/client/src/features/routing/routing-module.ts#L43).

**Comment by rsoika** [comment]
@tortmayr, Thanks, yes using the default routerkind works with the rebind command you showed me. The only reason for a custom routerkind was the idea to contribute my results later. But that we can discuss after I have some useful new implementation. 

Can you please explain why the `GLSPManhattanEdgeRouter` overwrite the method `applyInnerHandleMoves`? There are no comments in the code. And I wonder if this is a good place to do my own customization of computing new routing points?

And another problem I came across are the missing sprotty classes. For example if I would like to overwrite the route method, 

```javascript
  override route(edge: SRoutableElementImpl): RoutedPoint[] {
         return super.route(edge);
   }
```

how can I import the `SRoutableElementImpl` into my own implementation? I can't find out which library I need to import for this kind of classes.

**Reply by tortmayr** [reply, marked as answer]
>Can you please explain why the GLSPManhattanEdgeRouter overwrite the method applyInnerHandleMoves? There are no comments in the code. 

Looks like this was a workaround for https://github.com/eclipse-sprotty/sprotty/issues/310.
Since this is now fixed directly in sprotty, the override is no longer necessary. Thanks for pointing that out,
we will remove this override

>And I wonder if this is a good place to do my own customization of computing new routing points?
Unfortunately, I'm not that familiar with the details for the manhattan router or edge routers in general.
However, it seem to me like the routing points are computed in the `route` method:
```ts
 route(edge: SRoutableElementImpl): RoutedPoint[] {
        // ...
        const routedPoints: RoutedPoint[] = [];
        routedPoints.push({ kind: 'source', ...sourceAnchor});
        routedCorners.forEach(corner => routedPoints.push(corner));
        routedPoints.push({ kind: 'target', ...targetAnchor});
        return routedPoints;
    }
  ```
  
  So probably overriding this method is a good starting point.
  
  >And another problem I came across are the missing sprotty classes. For example if I would like to overwrite the route method,

  Sprotty Smodel classes are not missing but they are aliased to the GModel namespace in GLSP. 
  In sprotty model classes (since 1.0.0) use the following naming pattern: `S<ModelName>Impl`
  In glsp the corresponding aliased class use this pattern: `G<ModelName>`
  
  So in your case instead of using `SRoutableElementImpl` you have to use the `GRoutableElement`
```ts
  override route(edge: GRoutableElement): RoutedPoint[] {
         return super.route(edge);
   }
   ```

**Reply by tortmayr** [reply]
FYI: I opened a PR to remove the unnecessary `applyInnerHandleMoves` override: https://github.com/eclipse-glsp/glsp-client/pull/404
